File: go/tree/106_build_tree.go
package tree

import "slices"

// buildTree_ 从中序与后序遍历序列构造二叉树
func buildTree_(inorder []int, postorder []int) *TreeNode {

	// 输入：inorder = [9,3,15,20,7], postorder = [9,15,7,20,3]
	// 输出：[3,9,20,null,null,15,7]

	if len(inorder) == 0 {
		return nil
	}

	rootVal := postorder[len(postorder)-1]
	root := &TreeNode{rootVal, nil, nil}

	leftSize := slices.Index(inorder, rootVal)

	root.Left = buildTree_(inorder[:leftSize], postorder[:leftSize])
	root.Right = buildTree_(inorder[leftSize+1:], postorder[leftSize:len(postorder)-1])

	return root
}

// buildTree_2 从中序与后序遍历序列构造二叉树  迭代的方法实现
func buildTree_2(inorder []int, postorder []int) *TreeNode {

	// 输入：inorder = [9,3,15,20,7], postorder = [9,15,7,20,3]
	// 输出：[3,9,20,null,null,15,7]

	if len(inorder) == 0 {
		return nil
	}

	root := &TreeNode{postorder[len(postorder)-1], nil, nil}
	stack := make([]*TreeNode, 0)
	stack = append(stack, root)

	var inorderIndex int = len(inorder) - 1

	for i := len(postorder) - 2; i >= 0; i-- {

		val := postorder[i]
		node := stack[len(stack)-1]

		if node.Val != inorder[inorderIndex] {
			node.Right = &TreeNode{val, nil, nil}
			stack = append(stack, node.Right)
		} else {

			for len(stack) != 0 && stack[len(stack)-1].Val == inorder[inorderIndex] {
				node = stack[len(stack)-1]
				stack = stack[:len(stack)-1]
				inorderIndex--
			}

			node.Left = &TreeNode{val, nil, nil}
			stack = append(stack, node.Left)
		}

	}
	return root
}

File: rust/src/tree/_106_build_tree.rs
use super::TreeNode;
use std::cell::RefCell;
use std::rc::Rc;

/// 从中序与后序遍历序列构造二叉树
pub fn build_tree(inorder: Vec<i32>, postorder: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
    build(&inorder, &postorder)
}

fn build(inorder: &[i32], postorder: &[i32]) -> Option<Rc<RefCell<TreeNode>>> {
    if inorder.is_empty() {
        return None;
    }

    let root_val = postorder.last()?;
    let left_size = inorder.iter().position(|x| x == root_val)?;

    let root = Some(Rc::new(RefCell::new(TreeNode::new(*root_val))));

    if let Some(r) = &root {
        r.borrow_mut().left = build(&inorder[..left_size], &postorder[..left_size]);
        r.borrow_mut().right = build(
            &inorder[left_size + 1..],
            &postorder[left_size..postorder.len() - 1],
        );
    }
    return root;
}
